File: scripts/ie/offsets.py
import re
import sys
from collections import OrderedDict

from bs4 import BeautifulSoup
from markdown import markdown


def get_offset_prefix(file_version, data_file_name):  # eff_v2 / body.yml
    base = re.sub("_v.*", "", file_version)
    version = re.sub(".*_v", "", file_version)
    version = version.replace(".", "")
    if version == "1":
        version = ""

    # custom prefix for some data structures
    fbase = data_file_name.replace(".yml", "")
    fbase_map = {"header": "", "body": "", "extended_header": "head"}
    suffix = fbase_map.get(fbase, fbase)

    prefix = f"{base}{version}_"
    if suffix != "":
        prefix = prefix + f"{suffix}_"
    prefix = prefix.upper()
    return prefix


def get_offset_id(item, prefix):
    """
    Tries to get an id for an offset item
    """
    # custom IElib id
    if "id" in item:
        return prefix + item["id"]

    # no custom id, constructing from description
    iid = string_to_id(item["desc"], prefix)
    return iid


def string_to_id(line, prefix):
    """
    Tries to convert a string to an id suitable for a define
    """
    iid = line.lower()

    # strip links
    html = markdown(iid)
    iid = "".join(BeautifulSoup(html, features="lxml").findAll(text=True))

    # custom replacements
    iid = iid.replace("probability ", "probability")
    iid = iid.replace("usability ", "usability")
    iid = iid.replace("parameter ", "parameter")
    iid = iid.replace("resource ", "resource")
    iid = iid.replace("alternative", "alt")
    iid = iid.replace(".", "")

    iid = iid.replace(" ", "_")
    iid = prefix + iid

    # id must be alnum + '_' only
    if re.match(r"^[a-zA-Z0-9_]+$", iid):
        return iid
    # no good id found, aborting
    print(f'Bad id: "{iid}". Aborting.')
    sys.exit(1)


def get_offset_size(item):
    if "length" in item:
        return item["length"]
    size_map = {"byte": 1, "char": 1, "word": 2, "dword": 4, "resref": 8, "strref": 4}
    size = size_map[item["type"]]
    if "mult" in item:
        size = size * item["mult"]
    return size


def validate_offset(current_offset, item):
    """
    Check offset item against current offset, fail on mismatch.
    Mismatch probably means a bug in data.
    """
    if "offset" in item and item["offset"] != current_offset:
        print(f"Error: offset mismatch. Expected {current_offset}, got {item['offset']} for {item}")
        sys.exit(1)


def offset_is_unused(offset):
    """
    Checks if an IESDP offset item is unused

    @arg item - an offset loaded from IESDP _data/file_formats
    """
    if "unused" in offset or "unknown" in offset or ("desc" in offset and offset["desc"].lower() == "unknown"):
        return True
    return False


def offsets_to_definition(data, prefix):
    cur_off = 0
    if "offset" in data[0]:
        cur_off = data[0]["offset"]

    items = OrderedDict()
    for i in data:
        validate_offset(cur_off, i)

        size = get_offset_size(i)

        if offset_is_unused(i):
            cur_off += size
            continue

        iid = get_offset_id(i, prefix)
        items[iid] = hex(cur_off)
        cur_off += size
    return items

File: scripts/ie-update/src/ie/offsets.ts
/**
 * IESDP file format offset processing.
 * Handles prefix generation, offset ID construction, size calculation,
 * validation, and definition file generation from IESDP offset data.
 */

import type { OffsetItem } from "./types.js";

/** Map of offset type names to their byte sizes */
const SIZE_MAP: Readonly<Record<string, number>> = {
    byte: 1,
    char: 1,
    word: 2,
    dword: 4,
    resref: 8,
    strref: 4,
};

/** Custom replacements applied during string-to-id conversion */
const ID_REPLACEMENTS: ReadonlyArray<readonly [string, string]> = [
    ["probability ", "probability"],
    ["usability ", "usability"],
    ["parameter ", "parameter"],
    ["resource ", "resource"],
    ["alternative", "alt"],
    [".", ""],
];

/**
 * Generates an offset prefix from file version and data file name.
 * Example: ("itm_v2", "header.yml") -> "ITM_V2_"
 *          ("eff_v1", "body.yml") -> "EFF_"
 */
export function getOffsetPrefix(fileVersion: string, dataFileName: string): string {
    const base = fileVersion.replace(/_v.*/, "");
    let version = fileVersion.replace(/.*_v/, "");
    version = version.replaceAll(".", "");
    if (version === "1") {
        version = "";
    }

    const fbase = dataFileName.replace(".yml", "");
    const fbaseMap: Readonly<Record<string, string>> = {
        header: "",
        body: "",
        extended_header: "head",
    };
    // Safe: fbase checked via `in` operator above
    const suffix = fbase in fbaseMap ? fbaseMap[fbase]! : fbase;

    let prefix = `${base}${version}_`;
    if (suffix !== "") {
        prefix = `${prefix}${suffix}_`;
    }
    return prefix.toUpperCase();
}

/**
 * Simplifies file format version for IElib directory naming.
 * Example: "eff_v2" -> "eff2", "itm_v1" -> "itm"
 */
export function getFormatVersion(fileVersion: string): string {
    const base = fileVersion.replace(/_v.*/, "");
    let version = fileVersion.replace(/.*_v/, "");
    version = version.replaceAll(".", "");
    if (version === "1") {
        version = "";
    }
    return `${base}${version}`;
}

/**
 * Gets or generates an ID for an offset item.
 * Uses the item's custom 'id' field if present, otherwise constructs from description.
 */
export function getOffsetId(item: OffsetItem, prefix: string): string {
    if (item.id !== undefined) {
        return prefix + item.id;
    }
    return stringToId(item.desc, prefix);
}

/**
 * Converts a description string to a valid define identifier.
 * Strips markdown links, applies custom replacements, and validates the result.
 */
export function stringToId(line: string, prefix: string): string {
    let iid = line.toLowerCase();

    // Strip markdown links: [text](url) -> text
    iid = iid.replace(/\[([^\]]+)\]\([^)]+\)/g, "$1");
    // Strip any remaining HTML tags
    iid = iid.replace(/<[^>]+>/g, "");

    for (const [orig, repl] of ID_REPLACEMENTS) {
        iid = iid.replaceAll(orig, repl);
    }

    iid = iid.replaceAll(" ", "_");
    iid = prefix + iid;

    if (/^[a-zA-Z0-9_]+$/.test(iid)) {
        return iid;
    }

    throw new Error(`Bad id: "${iid}". Aborting.`);
}

/**
 * Calculates the byte size of an offset item.
 * Uses explicit 'length' if present, otherwise derives from type and multiplier.
 */
export function getOffsetSize(item: OffsetItem): number {
    if (item.length !== undefined) {
        return item.length;
    }
    const size = SIZE_MAP[item.type];
    if (size === undefined) {
        throw new Error(`Unknown offset type: ${item.type}`);
    }
    if (item.mult !== undefined) {
        return size * item.mult;
    }
    return size;
}

/**
 * Validates that the current tracked offset matches the item's declared offset.
 * Throws on mismatch, indicating a bug in the source data.
 */
export function validateOffset(currentOffset: number, item: OffsetItem): void {
    if (item.offset !== undefined && item.offset !== currentOffset) {
        throw new Error(
            `Offset mismatch. Expected ${currentOffset}, got ${item.offset} for ${JSON.stringify(item)}`
        );
    }
}

/**
 * Checks whether an offset item should be skipped (unused or unknown).
 */
export function offsetIsUnused(offset: OffsetItem): boolean {
    if (offset.unused !== undefined || offset.unknown !== undefined) {
        return true;
    }
    if (offset.desc.toLowerCase() === "unknown") {
        return true;
    }
    return false;
}

/**
 * Converts an array of offset items to a definition map of id -> hex value.
 * Skips unused/unknown items. Validates offsets along the way.
 */
export function offsetsToDefinition(
    data: readonly OffsetItem[],
    prefix: string
): Map<string, string> {
    const firstItem = data[0];
    if (firstItem === undefined) {
        return new Map();
    }

    let curOff = firstItem.offset ?? 0;
    const items = new Map<string, string>();

    for (const item of data) {
        validateOffset(curOff, item);
        const size = getOffsetSize(item);

        if (offsetIsUnused(item)) {
            curOff += size;
            continue;
        }

        const iid = getOffsetId(item, prefix);
        items.set(iid, `0x${curOff.toString(16)}`);
        curOff += size;
    }

    return items;
}
